gen_exports.py

```diff
@@ -0,0 +1,223 @@
+
+    
+"""
+
+TEMPLATE = """locals()[LOCALS_KEY_KI_PROTECTION_ENABLED] = True
+try:
+    return GLOBAL_RUN_CONTEXT.{}.{}
+except AttributeError:
+    raise RuntimeError('must be called from async context')
+"""
+
+
+def is_function(node):
+    """Check if the AST node is either a function
+    or an async function
+    """
+    return isinstance(node, ast.FunctionDef) \
+           or isinstance(node, ast.AsyncFunctionDef)
+
+
+def is_public(node):
+    """Check if the AST node has a _public decorator
+    """
+    return is_function(node) \
+           and node.decorator_list \
+           and isinstance(node.decorator_list[-1], ast.Name) \
+           and node.decorator_list[-1].id == '_public'
+
+
+def get_public_methods(tree):
+    """ Return a list of methods marked as public.
+    The function walks the given tree and extracts
+    all objects that are functions and have a
+    doc string that starts with PUBLIC
+    """
+    methods = []
+    for node in ast.walk(tree):
+        for child in ast.iter_child_nodes(node):
+            if is_public(child):
+                child.parent = node
+                child.module_file = tree.module_file
+                child.module_path = tree.module_path
+                methods.append(child)
+    return methods
+
+
+def get_module_trees_by_dir(source_dir):
+    """ Converts a list of modules into ast module objects
+    """
+    mod_files = astor.code_to_ast.find_py_files(source_dir)
+    modules = []
+    for mod_file in mod_files:
+        if not mod_file[1].startswith('_public'):
+            module = astor.code_to_ast.parse_file(os.path.join(*mod_file))
+            module.module_path = mod_file[0]
+            module.module_file = mod_file[1]
+            modules.append(module)
+    return modules
+
+
+def get_export_modules_by_dir(source_dir):
+    """Return all modules in the given directory
+    and its sub directories which contain methods
+    marked to be public as AST module trees
+    """
+    export_modules = []
+    modules = get_module_trees_by_dir(source_dir)
+
+    for module in modules:
+        methods = get_public_methods(module)
+        if methods:
+            export_modules.append(module)
+    return export_modules
+
+
+def get_doc_string(func):
+    """ Returns the doc string of a function
+    or an empty sting if none
+    """
+    doc = func.body[0]
+    if isinstance(doc, ast.Expr):
+        if hasattr(doc, 'value'):
+            if hasattr(doc.value, 's'):
+                return doc.value.s
+    return ""
+
+
+def create_passthrough_args(funcdef):
+    """ Create a pass through argument list
+    Every default argument value is
+    replaced by its argument corresponding
+    argument name
+    """
+    call_args = []
+    for arg in funcdef.args.args:
+        call_args.append(arg.arg)
+    for _, arg in zip(funcdef.args.defaults[::-1], funcdef.args.args[::-1]):
+        call_args.remove(arg.arg)
+        call_args.append(arg.arg + "=" + arg.arg)
```

Comment:
I thought so first too, but the tests failed without the additional default arguments. Especially _NO_SEND did not work without this little trick. I could try to remove it again and check if I made a mistake, but I spend quite some time to figure this out, so I know I have to repeat the default argument otherwise some tests fail.
What do you suggest?